File: 贪心/chapter5/5.1/No2645/main.cpp
#include <iostream>
#include <stack>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
  int addMinimum(string word) {
    stack<char> st;
    st.push('a');
    st.push('b');
    st.push('c');
    int i = 0, n = word.length();
    while (i < n) {
      int val = (st.top() - 96) % 3;
      if (val == word[i] - 'a') {
        st.push(word[i++]);
      } else {
        st.push((st.top() - 96) % 3 + 97);
      }
    }
    int cur = st.size() + 'c' - st.top();

    return cur - n - 3;
  }
};

signed main() {
  string word = "aaa";
  Solution sol;
  cout << sol.addMinimum(word) << endl;

  return 0;
}

File: 贪心/chapter5/5.1/No2645/main.py
class Solution:
    def addMinimum(self, word: str) -> int:
        st = []
        a = list(word)
        n = len(a)
        i = 0
        base = list("abc")
        for c in base:
            st.append(c)
        while i < n:
            if (ord(st[-1]) - 64) % 3 == ord(a[i]) - 65:
                st.append(a[i])
                i += 1
            else:
                st.append(chr((ord(st[-1]) - 64) % 3 + 65))
        return len(st)


def main():
    word = "aaa"
    sol = Solution()
    print(sol.addMinimum(word))


main()
